src/repos/ArticleRepo.ts

```diff
@@ -83,6 +84,81 @@ const getArticlesByPublicationSlugs = async (
     });
 };
 
+const getShuffledArticlesByPublicationSlugs = async (
+  slugs: string[],
+  limit: number = DEFAULT_LIMIT,
+  offset: number = DEFAULT_OFFSET,
+  timeRange = 12, // int value in months before current date to get articles from
+): Promise<Article[]> => {
+  // return a list of ArticleModels sorted by date, but each publisher has to show up at least once
+  // before articles by the same publisher shows up again
+  // e.g. publisher 1 has articles 1.1 (published jan 1), 1.2 (published jan 6)
+  //      publisher 2 has articles 2.1 (published jan 5), 2.2 (published jan 4)
+  //      publisher 3 has articles 3.1 (published jan 2), 3.2 (published jan 3)
+  // the function should return [1.1, 3.1, 2.1, 3.2, 2.2, 1.2]
+  const uniqueSlugs = [...new Set(slugs)];
+  const since = new Date(
+    new Date().getFullYear(),
+    new Date().getMonth() - timeRange,
+    new Date().getDate(),
+  );
+
+  let mostByOnePub = 0;
+
+  // by using the aggregate() function, create a dictionary of
+  // { publicationSlug: [ArticleModel] }
+  const articlesByPub = await ArticleModel.aggregate([
+    {
+      $match: {
+        'publication.slug': { $in: uniqueSlugs },
+        date: { $gte: since },
+      },
+    },
+    {
+      $sort: { date: -1 },
+    },
+    {
+      $skip: offset,
+    },
+    {
+      $limit: limit,
+    },
+    {
+      $group: {
+        _id: '$publication.slug',
+        articles: { $push: '$$ROOT' },
+      },
+    },
+  ]).then((articles) => {
+    const filteredArticles = {};
+    for (const a of articles) {
+      filteredArticles[a._id] = a.articles.filter(
+        (article) => article !== null && !isArticleFiltered(article),
+      );
+      mostByOnePub = Math.max(mostByOnePub, filteredArticles[a._id].length);
+    }
+    return filteredArticles;
+  });
```

Comment:
You should never be using `await` and `.then` together in an async function -- always stick to one or the other because they're different methods of resolving promises. Since you already `await` the results of the aggregation, you can just perform the for loop operation directly afterwards. Here's a rough outline of what this looks like:

```suggestion
  const articles = await ArticleModel.aggregate([
    {
      $match: {
        'publication.slug': { $in: uniqueSlugs },
        date: { $gte: since },
      },
    },
    {
      $sort: { date: -1 },
    },
    {
      $skip: offset,
    },
    {
      $limit: limit,
    },
    {
      $group: {
        _id: '$publication.slug',
        articles: { $push: '$$ROOT' },
      },
    },
  ])
  
    const articlesByPub = {};
    for (const a of articles) {
      filteredArticles[a._id] = a.articles.filter(
        (article) => article !== null && !isArticleFiltered(article),
      );
      mostByOnePub = Math.max(mostByOnePub, filteredArticles[a._id].length);
    }
```